### ISSUE TITLE
Basic PAM support hangs in v4.4.0-rc.3

### ISSUE THREAD

**Issue by benarent** [gravitational/teleport#4551]
### Description

**What happened**:

![image](https://user-images.githubusercontent.com/559288/96038460-045cdd00-0e1c-11eb-9e95-e0c791e99dec.png)
 
Config on Node
```yaml
teleport:
    auth_token: cxxx
    auth_servers:
     - teleport-4-4-fips.asteroid.earth:3080
auth_service:
    enabled: "false"
ssh_service:
    enabled: "true"
    enhanced_recording:
       enabled: "true"
    pam:
       enabled: "true"
       service_name: "sshd"
    enabled: "false"
```
#### On the IoT Node
```log
DEBU [NODE]      Handling request shell, want reply true. id:1 idle:30m0s local:172.31.14.24:38934 login:root remote:73.202.138.51:46191 teleportUser:ben@gravitational.com regular/sshserver.go:1291
INFO [NODE]      Creating (interactive) session 81d0c72c-c1de-4e4d-bf39-56b3cb076529. id:1 idle:30m0s local:172.31.14.24:38934 login:root remote:73.202.138.51:46191 teleportUser:ben@gravitational.com srv/sess.go:209
DEBU [SESSION:N] Using sync streamer for session 81d0c72c-c1de-4e4d-bf39-56b3cb076529. srv/sess.go:1035
INFO [SESSION:N] New party ServerContext(73.202.138.51:46191->172.31.14.24:38934, user=root, id=1) party(id=8e6b4f5e-24e3-4ca1-a2b1-3226cffe096c) joined session: 81d0c72c-c1de-4e4d-bf39-56b3cb076529 srv/sess.go:1225
DEBU             Will join session session(id=81d0c72c-c1de-4e4d-bf39-56b3cb076529, parties=1) for SSH connection 73.202.138.51:46191. srv/ctx.go:424
DEBU [SESSION:N] Starting poll and sync of terminal size to all parties. srv/sess.go:1156
DEBU [NODE]      Handling request window-change, want reply false. id:1 idle:30m0s local:172.31.14.24:38934 login:root remote:73.202.138.51:46191 teleportUser:ben@gravitational.com regular/sshserver.go:1291
DEBU             Will join session session(id=81d0c72c-c1de-4e4d-bf39-56b3cb076529, parties=1) for SSH connection 73.202.138.51:46191. srv/ctx.go:424
```

Logs once I cancel it from the UI.

```
DEBU [SSH:NODE]  Closed connection 73.202.138.51:46191. sshutils/server.go:440
DEBU             AgentServer(/tmp/teleport-460856269/teleport-3877.socket) is closing teleagent/agent.go:141
ERRO [NODE]      agent server for user "ben@gravitational.com" stopped: accept unix /tmp/teleport-460856269/teleport-3877.socket: use of closed network connection id:1 idle:30m0s local:172.31.14.24:38934 login:root remote:73.202.138.51:46191 teleportUser:ben@gravitational.com regular/sshserver.go:829
INFO [SESSION:N] Party member 8e6b4f5e-24e3-4ca1-a2b1-3226cffe096c left session 81d0c72c-c1de-4e4d-bf39-56b3cb076529. srv/sess.go:1232
DEBU [NODE]      Client 73.202.138.51:46191 disconnected. id:1 idle:30m0s local:172.31.14.24:38934 login:root remote:73.202.138.51:46191 teleportUser:ben@gravitational.com regular/sshserver.go:1255
INFO [SESSION:N] Closing party 8e6b4f5e-24e3-4ca1-a2b1-3226cffe096c srv/sess.go:1383
WARN [SESSION:N] Unable to send session.join to 73.202.138.51:46191: EOF. srv/sess.go:278
INFO [NODE]      Removing party ServerContext(73.202.138.51:46191->172.31.14.24:38934, user=root, id=1) party(id=8e6b4f5e-24e3-4ca1-a2b1-3226cffe096c) from session 81d0c72c-c1de-4e4d-bf39-56b3cb076529 id:1 idle:30m0s local:172.31.14.24:38934 login:root remote:73.202.138.51:46191 teleportUser:ben@gravitational.com srv/sess.go:1080
INFO [AUDIT]     session.data addr.remote:73.202.138.51:46191 code:T2006I ei:2.147483646e+09 event:session.data login:root namespace:default rx:3790 server_id:9042e144-670d-4bf7-a4be-81487d142d62 sid:81d0c72c-c1de-4e4d-bf39-56b3cb076529 time:2020-10-14T19:52:34.448Z tx:4631 uid:44e21444-552d-45e9-82b4-30d33e14fbd3 user:ben@gravitational.com events/emitter.go:202
DEBU [NODE]      Releasing associated resources - context has been closed. id:1 idle:30m0s local:172.31.14.24:38934 login:root remote:73.202.138.51:46191 teleportUser:ben@gravitational.com srv/monitor.go:208
```

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

### Environment
- Teleport version (use `teleport version`):
- Tsh version (use `tsh version`):
- OS (e.g. from `/etc/os-release`):

- Where are you running Teleport? (e.g. AWS, GCP, Dedicated Hardware):

**Browser environment**

- Browser Version (for UI-related issues):
- Install tools:
- Others:

**Relevant Debug Logs If Applicable**

- `tsh --debug`
- `teleport --debug`

**Comment by awly** [comment]
@benarent what do you have in `/etc/pam.d/sshd`?
There may be some `auth` module that waits for user input or something similar

**Comment by awly** [comment]
Dumping the current investigation state, will continue tomorrow:
- issue introduced in https://github.com/gravitational/teleport/pull/3966 - teleport triggering `pam_authenticate` on every connection
- the default `/etc/pam.d/sshd` calls `pam_unix.so` under the `auth` section
  - `pam_unix.so` does local user password validation
- so `tsh` hangs on that password prompt
  - not clear why the actual prompt isn't displayed
  - entering a newline unblocks it, but password validation always fails, even with a valid password
- I injected debug steps into `/etc/pam.d/sshd` and it looks like OpenSSH doesn't call `pam_authenticate` if a different authentication method (e.g. public key) has succeeded

Next steps:
- confirm that OpenSSH code skips `pam_authenticate`
  - find clues on why
- since our motivation for using `pam_authenticate` is or 2FA (e.g. Duo), see if we can inject it through `account` or `session` PAM stage instead

**Comment by awly** [comment]
One possible "hack" to unblock 4.4 - if `pam.service_name` is `sshd` in `teleport.yaml`, skip the `pam_authenticate` step.
It's gross, but keeps Duo 2FA support while not breaking people re-using the `sshd` config.
@benarent WDYT?

**Comment by benarent** [comment]
Another interesting note, when trying the PAM Duo plugin. I did make extra changes to SSH config. https://duo.com/docs/duounix#public-key-authentication. 

The other option we recommend to users to use `pam_exec.so` to create users. If we do go with the above proposal, we should check that this doesn't break this. https://gravitational.com/teleport/docs/features/ssh-pam/#examples

**Comment by awly** [comment]
Thanks @benarent, the Duo docs provided the necessary hint: `AuthenticationMethods` config field.

**TL;DR:** we need a config flag.

### boring techincal details

OpenSSH has a set of authentication methods: "gssapi-with-mic", "hostbased", "keyboard-interactive", "none", "password" and "publickey". All of these are gated in `sshd_config` with flags like `PasswordAuthentication no` or `PubkeyAuthentication yes`. 

These flags only specify whether a specific method is _supported_, but by default any one of them is sufficient to authenticate.
There is another flag, `AuthenticationMethods` which specifies a list of methods that **must** succeeds before authentication completes. So, for example `AuthenticationMethods: publickey,password` will only allow users in after **both** key-based and password auth was completed (even if any other methods succeed too).

"keyboard-interactive" method means `pam_authenticate` (assuming `UsePAM yes` is set).
So, with default `sshd_config`, `pam_authenticate` is only triggered if public key, GSSAPI and password auth all fail.

For Teleport, since we _always_ require cert auth to succeed, calling `pam_authenticate` always is like `AuthenticationMethods: publickey,keyboard-interactive` in `sshd_config`.
Instead, we need a field in `teleport.yaml` for users to opt into that, same as `sshd_config` does.

#### example `sshd_config`s.

Public key auth OR password auth OR `pam_authenticate`, whichever succeeds first.
`pam_authenticate` not triggered if pubkey succeeds.
```
UsePAM yes
KbdInteractiveAuthentication yes
PubkeyAuthentication yes
PasswordAuthentication yes
```

Public key auth required, `pam_authenticate` not triggered.
```
UsePAM yes
KbdInteractiveAuthentication yes
PubkeyAuthentication yes
PasswordAuthentication yes
AuthenticationMethods publickey
```

Both public key auth and `pam_authenticate` required (this is what Duo docs recommend)
```
UsePAM yes
KbdInteractiveAuthentication yes
PubkeyAuthentication yes
PasswordAuthentication no
AuthenticationMethods publickey,keyboard-interactive
```

**Comment by benarent** [comment]
Thanks for explaining @awly So for Teleport. We would have something like this?

```
    # configures PAM integration. see below for more details.
    pam:
        enabled: yes
        # pam_authenticate defaults to no, for backwards compatibly .
        # if users want to use sshd, they would also need to update `sshd_configs` 
        pam_authenticate: yes
        service_name: teleport
```

**Comment by awly** [comment]
Yep. I'd call it `use_pam_auth` (`auth` is what you use under `/etc/pam.d/`, `pam_authenticate` is a C library call). Is that OK?
